kubemarine/patches/p1_upgrade_containerdio_version.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2021-2022 NetCracker Technology Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from kubemarine.core import flow
+from kubemarine.core.action import Action
+from kubemarine.core.patch import Patch
+from kubemarine.core.resources import DynamicResources
+from kubemarine import cri
+from kubemarine import packages
+
+
+class TheAction(Action):
+    def __init__(self):
+        super().__init__("Upgrade containerd.io version")
+
+    def run(self, res: DynamicResources):
+        cluster = res.cluster()
+
+        # check the OS family
+        if cluster.get_os_family() == 'debian':
+            # check CRI
+            if cluster.inventory['services']['cri']['containerRuntime']:
+                # get installed version
+                group = cluster.nodes['all'].get_accessible_nodes()
+                result = packages.detect_installed_packages_version_groups(group, 'containerd.io')
+                version = list(result['containerd.io'])[0].split('=')[1]
+                if '.'.join(version.split('.')[:2]) == '1.4':
```

Comment:
This is a random behavior in case different nodes have different versions. Probably make it more robust and raise error / log warning / install anyway in such case.

Reply:
This random behavior inherits from the package version detection method which was not perfectly designed.

Reply:
Could you specify problems in the package version detection method? For me, it does its task. It returns all versions from all nodes, including the case when the versions are different.

Reply:
Actually current implementation has one more problem. docker might be not installed on balancers, but you query balancers. Since you take the random result, it might be resolved to "not installed"

Reply:
The `cluster.yaml` structure implies only one version of any package for all nodes in the cluster. Therefore several version for the same package must raise an exception.

Reply:
> Therefore several versions for the same package must raise an exception.

Could you clarify, is this related to the design of detect_installed_packages_version_groups? The method is used in particular by check_paas. So processing of the result is delegated to the caller by design.
If you are about my initial proposal in this thread, then agree, raising of error might be one of the solutions. Though, you might additionally check if cache_package=False is configured, because it allows different package versions in the cluster.

Reply:
The `cache_package=False` option is just a bypass. I'm talking about a comprehensive solution which doesn't exist in a core, but you expect it in a patch.

Reply:
Comprehensive solution exists in core. All users of the method detect_installed_packages_version_groups take into account probably different package versions. If you need special behavior only for single patch and do not want to implement it directly in the patch, you can implement it in some other place.